src/App.tsx

```diff
@@ -1,24 +1,176 @@
+      break;
+    default:
+      return filteredTodos;
+  }
+
+  return filteredTodos;
+};
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filterStatus, setFilterStatus] = useState<Filter>(Filter.All);
+  const [errorType, setErrorType] = useState<Error | null>(null);
+  const [isLoading, setIsLoading] = useState(false);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [processingTodoIds, setProcessingTodoIds] = useState<number[]>([]);
+  const [todoTitle, setTodoTitle] = useState('');
+
+  useEffect(() => {
+    todoService.getTodos(USER_ID)
+      .then(response => {
+        setTodos(response);
+      });
+  }, []);
+
+  const handleDelete = (id: number) => {
+    setProcessingTodoIds(prevIds => [...prevIds, id]);
+    todoService.deleteTodo(id)
+      .then(() => {
+        setTimeout(() => {
+          setTodos(currentTodos => currentTodos.filter(post => post.id !== id));
+        }, 500);
+      })
+      .finally(() => setTimeout(() => {
+        setProcessingTodoIds(prevIds => [...prevIds]
+          .filter(prevId => prevId !== id));
+      }, 500));
+  };
+
+  const addTodo = (title: string) => {
+    setIsLoading(true);
+    setProcessingTodoIds(prevIds => [...prevIds, 0]);
+    setTempTodo({
+      id: 0,
+      title,
+      completed: false,
+      userId: USER_ID,
+    });
+    todoService.addTodo({
+      title,
+      completed: false,
+      userId: USER_ID,
+    })
+      .then(newTodo => {
+        setTodoTitle('');
+        setTimeout(() => {
+          setTodos(currentTodos => {
+            return [...currentTodos, newTodo];
+          });
+          setTempTodo(null);
+        }, 500);
+      })
+      .finally(() => {
+        setIsLoading(false);
+        setProcessingTodoIds(prevIds => [...prevIds]
+          .filter(prevId => prevId !== 0));
+      });
+  };
+
+  const updateTodo = (updatedTodo: Todo) => {
+    setProcessingTodoIds(currentTodos => [...currentTodos, updatedTodo.id]);
+
+    todoService.updateTodo(updatedTodo)
+      .then(() => setTodos(prev => (
+        prev.map(prevTodo => (
+          prevTodo.id === updatedTodo.id
+            ? updatedTodo
+            : prevTodo
+        ))
+      )))
+      .catch(() => setErrorType(Error.Update))
+      .finally(() => {
+        setProcessingTodoIds(currentTodos => currentTodos
+          .filter(todoId => updatedTodo.id !== todoId));
+      });
+  };
+
+  const toggleAll = async () => {
+    const isAllCompleted = todos.every(todo => todo.completed);
+
+    const todosToUpdate = todos.filter(todo => (isAllCompleted
+      ? todo.completed
+      : !todo.completed
+    ));
+
+    await Promise.all(todosToUpdate.map(todo => (
+      updateTodo({
+        ...todo,
+        completed: !isAllCompleted,
+      })
+    )));
+  };
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const isThereCompleted = todos.some(todo => todo.completed);
```

Comment:
you are using this in only one component, it is better to create this constant directly in todoFooter and not pass it through props

Reply:
Good point!